TestIs2WavAudioPlayer/I2sWavPlayer.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Device.I2s;
+using System.Diagnostics;
+using System.IO;
+using nanoFramework.Hardware.Esp32;
+
+namespace AudioPlayer
+{
+    public class I2sWavPlayer : IDisposable
+    {
+        public enum Bus
+        {
+            One = 1,
+            Two = 2
+        }
+
+        private readonly I2sDevice _i2S;
+        private readonly FileStream _stream;
+
+        public I2sWavPlayer(Bus bus, string audioFile, int bckPin = 32, int dataPin = 33, int wsPin = 25)
+        {
+            switch (bus)
+            {
+                case Bus.One:
+                    // I2S Audio device:
+                    Configuration.SetPinFunction(bckPin, DeviceFunction.I2S1_BCK);
+                    Configuration.SetPinFunction(dataPin, DeviceFunction.I2S1_DATA_OUT);
+                    Configuration.SetPinFunction(wsPin, DeviceFunction.I2S1_WS);
+                    //Configuration.SetPinFunction(5, DeviceFunction.I2S1_MCK);
+                    break;
+                case Bus.Two:
+                    // I2S Audio device:
+                    Configuration.SetPinFunction(bckPin, DeviceFunction.I2S2_BCK);
+                    Configuration.SetPinFunction(dataPin, DeviceFunction.I2S2_DATA_OUT);
+                    Configuration.SetPinFunction(wsPin, DeviceFunction.I2S2_WS);
+                    //Configuration.SetPinFunction(5, DeviceFunction.I2S2_MCK);
+                    break;
+            }
+
+            _stream = new FileStream(audioFile, FileMode.Open, FileAccess.Read);
+
+            var header = new byte[44];
+            var len = _stream.Read(header, 0, header.Length);
+            if (len != 44) throw new IOException("Not enough bytes in the wav file header.");
+
+            var headerParser = new WavFileHeader(header);
+
+            _i2S = new I2sDevice(new I2sConnectionSettings(1)
+            {
+                Mode = I2sMode.Master | I2sMode.Tx,
+                CommunicationFormat = I2sCommunicationFormat.I2S,
+
+                SampleRate = headerParser.SampleRate,
+                BitsPerSample = ToBitsPerSample(headerParser.BitsPerSample),
+                ChannelFormat = ToChannelFormat(headerParser.NumberOfChannels),
+                BufferSize = 40000
+            });
+        }
+
+        public void Dispose()
+        {
+            _i2S.Dispose();
+
+            _stream.Close();
+            _stream.Dispose();
+        }
+
+        public void Play()
+        {
+            _stream.Seek(44, SeekOrigin.Begin);
+
+            var buffer = new byte[10000];
+            try
+            {
+                while (true)
+                {
+                    var len = _stream.Read(buffer, 0, buffer.Length);
```

Comment:
you can have the definition of len outside of the loop. Same for SpanByte so it will not allocate new elements every time you need them

Reply:
I agree with len.
Considering SpanByte can potentially have a different length (probably only two different ones - full or partly full) on each iteration, I'd have to use Splice() at the end anyway, which creates a new SpanByte too, right? (maybe I could add a check for len != buffer.Length and only Splice then)

What do we care more about? some CPU cycles or RAM usage and GC pressure (which also includes CPU cycles). Hm.. I assume I should add the check :)